packages/dcos-integration-test/extra/test_units.py

```diff
@@ -0,0 +1,52 @@
+import glob
+import subprocess
+
+import pytest
+
+
+def test_verify_units():
+    """Test that all systemd units are valid."""
+    def _check_units(path):
+        """Verify all the units given by `path'"""
+        for file in glob.glob(path):
+            cmd = subprocess.run(
+                ["/usr/bin/systemd-analyze", "verify", "--no-pager", file],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT)
+            # systemd-analyze returns 0 even if there were warnings, so we
+            # assert that the command output was empty.
+            if cmd.stdout:
+                # We specifically allow directives that exist in newer systemd
+                # versions but will cause older systemd versions to complain.
+                # The "old" systemd version we are using as a baseline is
+                # systemd 219, which ships with CentOS 7.2.1511.
+                def _check_line(line):
+                    # `systemd-analyze verify` checks for errors in the given
+                    # unit files, as well as other files that are loaded
+                    # transitively. We do not want our tests to fail when
+                    # third-party software ships bad unit files, so we
+                    # explicitly check that 'dcos-' is present on a
+                    # line before checking if it is valid.
+                    if "dcos-" not in line:
+                        return True
+                    # The TasksMax directive exists in newer versions of systemd
+                    # where it is important to set. As we want to support multiple
+                    # versions of systemd our tests must ignore errors that
+                    # complain that it is an unknown directive.
+                    ignore_new_directives = ["TasksMax"]
```

Comment:
This would be great but it's tricky. The first idea that jumps to mind is to use the systemd version reported by `systemctl --version`. Unfortunately that reports `v219` in both CentOS 7.2 and 7.3, even though 7.3 has several useful features backported to it that 7.2's systemd does not have.

I get the feeling that to distinguish we'd have to use something like `systemd-analyze verify` and run into a chicken and egg problem.

If we shipped man pages one could grep those for mention of the ignored directives. Can you think of something similar, perhaps?